fish-rust/src/builtins/emit.rs

```diff
@@ -0,0 +1,51 @@
+use libc::c_int;
+
+use super::shared::{
+    builtin_print_help, io_streams_t, HelpOnlyCmdOpts, STATUS_CMD_OK, STATUS_INVALID_ARGS,
+};
+use crate::ffi::{self, parser_t, Repin};
+use crate::wchar_ffi::{wstr, W0String, WCharToFFI, WString};
+
+pub fn emit(
+    parser: &mut parser_t,
+    streams: &mut io_streams_t,
+    argv: &mut [&wstr],
+) -> Option<c_int> {
+    let cmd = argv[0];
+
+    let opts = match HelpOnlyCmdOpts::parse(argv, parser, streams) {
+        Ok(opts) => opts,
+        Err(err @ Some(_)) if err != STATUS_CMD_OK => return err,
+        Err(err) => panic!("Illogical exit code from parse_options(): {err:?}"),
+    };
+
+    if opts.print_help {
+        builtin_print_help(parser, streams, cmd);
+        return STATUS_CMD_OK;
+    }
+
+    let Some(event_name) = argv.get(opts.optind) else {
+        streams.err.append(WString::from_str(&format!("{cmd}: expected event name\n")));
+        return STATUS_INVALID_ARGS;
+    };
+
+    let event_args: Vec<W0String> = argv[opts.optind + 1..]
+        .iter()
+        .map(|s| W0String::from_ustr(s).unwrap())
+        .collect();
+    let event_arg_ptrs: Vec<ffi::wcharz_t> = event_args
+        .iter()
+        .map(|s| ffi::wcharz_t {
+            str_: s.as_ucstr().as_ptr().cast(),
```

Comment:
Correct. The W0String objects live until the function ends, so it's okay to pass their pointers.

I think we can simplify this, it seems `as_ucstr()` and `cast()` are unnecessary.
Also we can implement `From<&W0String> for wcharz_t`, in case we want to use this more often.
(could also implement `From<&Vec<W0String>> for Vec<wcharz_t>`).
Of course the conversion functions can be misused - it doesn't warn you if the `W0String` goes out of scope, which can  leave the `wcharz_t` as dangling reference.  I think that's acceptable.

```
diff --git a/fish-rust/src/builtins/emit.rs b/fish-rust/src/builtins/emit.rs
index bb4a5c99f..112eff5e1 100644
--- a/fish-rust/src/builtins/emit.rs
+++ b/fish-rust/src/builtins/emit.rs
@@ -38,8 +38,3 @@ pub fn emit(
         .collect();
-    let event_arg_ptrs: Vec<ffi::wcharz_t> = event_args
-        .iter()
-        .map(|s| ffi::wcharz_t {
-            str_: s.as_ucstr().as_ptr().cast(),
-        })
-        .collect();
+    let event_arg_ptrs: Vec<ffi::wcharz_t> = event_args.iter().map(ffi::wcharz_t::from).collect();
 
diff --git a/fish-rust/src/wchar_ffi.rs b/fish-rust/src/wchar_ffi.rs
index cc00c1ea7..14844a567 100644
--- a/fish-rust/src/wchar_ffi.rs
+++ b/fish-rust/src/wchar_ffi.rs
@@ -45,2 +45,8 @@ impl From<&wcharz_t> for WString {
 
+impl From<&W0String> for wcharz_t {
+    fn from(s: &W0String) -> Self {
+        wcharz_t { str_: s.as_ptr() }
+    }
+}
+
 /// Convert a wstr or WString to a W0String, which contains a nul-terminator.
```

Reply:
> Also we can implement From<&W0String> for wcharz_t, in case we want to use this more often.
 (could also implement From<&Vec<W0String>> for Vec<wcharz_t>).
> Of course the conversion functions can be misused - it doesn't warn you if the W0String goes out of scope, which can leave the wcharz_t as dangling reference.

I considered implementing `From`, but decided against it because it's too easy to cause gnarly UB with it.

Reply:
Right. We don't want to make it too easy.
If we use this pattern more often we can extract an `unsafe` function.
One safe alternative would be to not pass `wcharz_t` to C++ but `std::unique_ptr<wcstring>` which can be created with `.to_ffi()`.